tl

```diff
@@ -391,6 +411,109 @@ for _, include in ipairs(tlconfig["include_dir"]) do
+         else
+            return b
+         end
+      end)
+      :gsub(path_separator .. "+", path_separator))
+      :gsub(path_separator .. "+$", "")
+end
+
+local function path_concat(...)
+   local path = {}
+   for i = 1, select("#", ...) do
+      local fname = cleanup_file_name((select(i, ...)))
+      if #fname > 0 then
+         table.insert(path, fname)
+      end
+   end
+   return table.concat(path, path_separator)
+end
+
+local function remove_leading_path(leading_part, path)
+   local s, e = path:find("^" .. leading_part .. path_separator .. "?")
+   if s then
+      return path:sub(e+1, -1)
+   end
+   return path
+end
+
+local function traverse(dirname, emptyref,is_generated,generated_ref)
+   local files = {}
+   local paths = {} --lookup table for string paths to help
+   -- with pattern matching while iterating over a project
+   -- paths[files.foo.bar] -> "foo/bar"
+   emptyref = emptyref or {}
+   generated_ref = generated_ref or {}
+   for file in lfs.dir(dirname) do
+      if file ~= "." and file ~= ".." then
+         if lfs.attributes(path_concat(dirname, file), "mode") == "directory" then
+            local p
+            local prefix = dirname
+            files[file], p = traverse(path_concat(dirname, file), emptyref,is_generated,generated_ref)
+            paths[files[file]] = file
+            for k, v in pairs(p) do
+               paths[k] = path_concat(file, v)
+            end
+         else
+            -- storing a special entry in this table to it mark as empty could
+            -- interfere with convoluted or maliciously constructed directory
+            -- names so we use a table with specific metatable to mark
+            -- something as the end of a traversal to have a property attached
+            -- to the table, without creating an entry in the table
+            local meta_table = {empty=emptyref}
+            if is_generated then
+               meta_table["generated"] = generated_ref
+            end
+            files[file] = setmetatable({}, meta_table)
+            paths[files[file]] = file
+         end
+      end
+   end
+   return files, paths, emptyref
+end
+
+local function delete_dir(dir)
+   for file in lfs.dir(dir) do
+      if file ~= "." and file ~= ".." then 
+         local file_path = path_concat(dir,file)
+         if file ~= "." and file ~= ".." then
+            if lfs.attributes(file_path, 'mode') == 'file' then
+                  os.remove(file_path)
+            elseif lfs.attributes(file_path, 'mode') == 'directory' then
+                  delete_dir(file_path)
+            end
+         end
+      end
+   end
+   lfs.rmdir(dir)
+end
+
+local build_path
+
+local function cleanup_generated_files()
+   if build_path then
+      delete_dir(build_path)
+   end
+end
+if tlconfig["build_file"] then
+   local tmp_dir = os.getenv("TMP") or os.getenv("TMPDIR") or "/tmp" --fall back to /tmp, apparently some (mine?) computers don't have a variable set for it.
```

Comment:
Do we even need to do things in `/tmp/`? Can we not stay within the project file in `./generated_code`?

Reply:
I guess I can make a folder in the project directory where the generated teal files are placed, this would make it easier to cache the results. However I fear that this will look a lot uglier from the user side of things, as teal suddenly will generate a folder outside of whatever directory was chosen as the `build_dir` and drop files there.

I'm also not sure how that would interact if the user also has a directory with that name if not every .lua file in the corresponding folder is/can be generated. (I hope this sentence makes sense...)

Reply:
Perhaps something like `build_dir .. `/tl_build_generated` instead? This is an actually temporary directory which gets created then deleted, right?

Reply:
That could work as well, and you are correct in that it is a directory that gets created and later deleted.

However, why not use the tmp folder for temporary files? Is there something that I'm missing that makes not using the tmp folder preferred?

Also, eventually it is a good idea to only rerun the build file if it needs to which means that the user would see the `tl_build_generated` folder if we stick it in `build_dir` or anywhere else in the project, which doesn't sound nice if you ask me.

Unless we want to claim a folder for any build artifacts so that the files needed for https://github.com/teal-language/tl/issues/302 can be stored there as well?

In that case, I would recommend to make this folder changeable in tlconfig.lua and emit a warning (for now) if a build script is detected but this value is not set to make users aware of the fact that teal will "pollute" their project with files for internal use only. Once #302 is implemented this warning can either get lifted all together or get always emitted.